Touchscreen tt21100 no longer working after 2023.12.3 update

### The problem

Upgraded from 2023.11.6 to 2023.12.3 on my ESP32-S3-BOX (the old one) 
My touchscreen does not work - the screen or the home button directly below.  
Reverted back to 2023.11.6 and it started working again.  I know there is a breaking change for the touchscreen component as listed [here](https://github.com/esphome/esphome/pull/4596) but it doesn't look like I need to make any changes from my config?
Nothing fancy going on for the box, just using it as a display with some touchscreen buttons to turn office lights on/off.

### Which version of ESPHome has the issue?

2023.12.3

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2023.12.3

### What platform are you using?

ESP32

### Board

ESP32-S3-BOX

### Component causing the issue

touchscreen

### Example YAML snippet

```yaml
---
esphome:
  name: flexbox
  friendly_name: ESP32 S3 Box
  platformio_options:
    board_build.flash_mode: dio
  project:
    name: esphome.voice-assistant
    version: "1.0"
  min_version: 2023.10.5
  includes:
    - common.h

esp32:
  board: esp32s3box
  framework:
    type: esp-idf
    sdkconfig_options:
      CONFIG_ESP32S3_DEFAULT_CPU_FREQ_240: "y"
      CONFIG_ESP32S3_DATA_CACHE_64KB: "y"
      CONFIG_ESP32S3_DATA_CACHE_LINE_64B: "y"

logger:
  hardware_uart: USB_SERIAL_JTAG
  level: DEBUG
  logs:
    component: ERROR

api: 
  encryption:
    key: 3IzrqZs4csVkrt7/lmr238Z+HRP/vAZCbgqGCOe8zsQ=

ota:
  password: !secret ota_password


dashboard_import:
  package_import_url: github://esphome/firmware/voice-assistant/esp32-s3-box.yaml@main

packages:
  <<: !include_dir_named common  

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  fast_connect: True
  manual_ip:
    static_ip: 192.168.25.74
    gateway: 192.168.25.254
    subnet: 255.255.255.0
    dns1: 192.168.1.210

binary_sensor:
  - platform: tt21100
    name: Home-Button
    index: 0
    tt21100_id: my_touchscreen
    internal: False
  - platform: touchscreen
    name: "FanLo"
    x_min: 0
    x_max: 105
    y_min: 0
    y_max: 80
  - platform: touchscreen
    name: "FanMed"
    x_min: 106
    x_max: 210
    y_min: 0
    y_max: 80
  - platform: touchscreen
    name: "FanHi"
    x_min: 211
    x_max: 320
    y_min: 0
    y_max: 80
  - platform: touchscreen
    name: "FanOff"
    x_min: 106
    x_max: 210
    y_min: 81
    y_max: 160
  - platform: touchscreen
    name: "Lamp"
    x_min: 40
    x_max: 140
    y_min: 161
    y_max: 240
  - platform: touchscreen
    name: "FanLight"
    x_min: 180
    x_max: 280
    y_min: 161
    y_max: 240



output:
  - platform: ledc
    pin: GPIO45
    id: backlight_output

light:
  - platform: monochromatic
    output: backlight_output
    name: LCD Backlight
    id: led
    restore_mode: ALWAYS_ON
    disabled_by_default: true
    default_transition_length: 0s

#esp_adf:

#microphone:
#  - platform: esp_adf
#    id: box_mic

#speaker:
#  - platform: esp_adf
#    id: box_speaker

#external_components:
#  - source: github://pr#5230
#    components: esp_adf
#    refresh: 0s

spi:
  clk_pin: GPIO7
  mosi_pin: GPIO6

# some HW Components of the S3-Box
i2c:
  scl: GPIO18
  sda: GPIO8
  scan: true

touchscreen:
  platform: tt21100
  id: my_touchscreen
  # address: 0x24
  interrupt_pin: GPIO3

# source: https://gist.github.com/BrettSheleski/5a33f6880c5ad4033ac2dc4cb5797ac1
# material UI icons - from https://pictogrammers.github.io/@mdi/font/5.3.45/
font:
  - file: "gfonts://Roboto@500"
    id: font_large
    size: 70
    glyphs: "0123456789:APM."
  - file: "gfonts://Roboto@500"
    id: font_medium
    size: 30
  - file: "gfonts://Roboto@500"
    id: font_small
    size: 15
  - file: "Arial.ttf"
    id: font_arial_big
    size: 70
  - file: "materialdesignicons-webfont.ttf"
    id: font_icons
    size: 130
    glyphs:
      - "\U000F0594" # clear-night
      - "\U000F0590" # cloudy
      - "\U000F0591" # fog
      - "\U000F0592" # hail
      - "\U000F0593" # lightning
      - "\U000F067E" # lightning-rainy
      - "\U000F0F31" # night-partly-cloudy
      - "\U000F0595" # partly-cloudy
      - "\U000F0596" # pouring
      - "\U000F0597" # rainy
      - "\U000F0F36" # snowy
      - "\U000F067F" # snowy-rainy
      - "\U000F0599" # sunny
      - "\U000F059D" # windy
      - "\U000F059E" # windy-variant
      - "\U000F0F38" # exCentereptional
      # Garage and others Status
      - "\U000F06D9" # garage closed
      - "\U000F06DA" # garage opened

       
  - file: "materialdesignicons-webfont.ttf"
    id: font_icons_med
    size: 45     
    glyphs:
      # Small Weather Icons
      - "\U000F0594" # clear-night
      - "\U000F0590" # cloudy
      - "\U000F0591" # fog
      - "\U000F0592" # hail
      - "\U000F0593" # lightning
      - "\U000F067E" # lightning-rainy
      - "\U000F0F31" # night-partly-cloudy
      - "\U000F0595" # partly-cloudy
      - "\U000F0596" # pouring
      - "\U000F0597" # rainy
      - "\U000F0F36" # snowy
      - "\U000F067F" # snowy-rainy
      - "\U000F0599" # sunny
      - "\U000F059D" # windy
      - "\U000F059E" # windy-variant/wind avg
      - "\U000F0F38" # exCentereptional
      # Misc icons
      - "\U000F10C2" # Temperature High
      - "\U000F10C3" # Temperature Low
      - "\U000F054B" # umbrella
      - "\U000F02E3" # Bed
      - "\U000F0E2E" # Fireplace On
      - "\U000F0E2F" # Fireplace Off
      - "\U000F04B9" # sofa
      - "\U000F023F" # wind max
      - "\U000F140B" # lightning
      - "\U000F028F" # fridge
      - "\U000F11B1" # fridge problem
      - "\U000F1435" # no motion
      - "\U000F0D91" # motion detected
      - "\U000F17FF" # UV Sun
      - "\U000F059C" # Sunrise
      - "\U000F059B" # Sunset
      # moon phases
      - "\U000F0F61" # moon-first-quarter
      - "\U000F0F62" # moon-full
      - "\U000F0F63" # moon-last-quarter
      - "\U000F0F64" # moon-new
      - "\U000F0F65" # moon-waning-crescent
      - "\U000F0F66" # moon-waning-gibbous
      - "\U000F0F67" # moon-waxing-crescent
      - "\U000F0F68" # moon-waxing-gibbous
      # fan and light
      - "\U000F1472" # Fan speed 1
      - "\U000F1473" # Fan speed 2 
      - "\U000F1474" # Fan speed 3
      - "\U000F081D" # Fan off
      - "\U000F06B5" # Lamp
      - "\U000F0769" # Ceiling Light       

  - file: "materialdesignicons-webfont.ttf"
    id: font_icons_small
    size: 15
    glyphs:
      # Small Weather Icons
      - "\U000F0594" # clear-night
      - "\U000F0590" # cloudy
      - "\U000F0591" # fog
      - "\U000F0592" # hail
      - "\U000F0593" # lightning
      - "\U000F067E" # lightning-rainy
      - "\U000F0F31" # night-partly-cloudy
      - "\U000F0595" # partly-cloudy
      - "\U000F0596" # pouring
      - "\U000F0597" # rainy
      - "\U000F0F36" # snowy
      - "\U000F067F" # snowy-rainy
      - "\U000F0599" # sunny
      - "\U000F059D" # windy
      - "\U000F059E" # windy-variant/wind avg
      - "\U000F0F38" # exCentereptional
      # Misc icons
      - "\U000F10C2" # Temperature High
      - "\U000F10C3" # Temperature Low
      - "\U000F054B" # umbrella
      - "\U000F02E3" # Bed
      - "\U000F0E2E" # Fireplace On
      - "\U000F0E2F" # Fireplace Off
      - "\U000F04B9" # sofa
      - "\U000F023F" # wind max
      - "\U000F140B" # lightning
      - "\U000F028F" # fridge
      - "\U000F11B1" # fridge problem
      - "\U000F1435" # no motion
      - "\U000F0D91" # motion detected
      - "\U000F17FF" # UV Sun
      - "\U000F059C" # Sunrise
      - "\U000F059B" # Sunset
      # moon phases
      - "\U000F0F61" # moon-first-quarter
      - "\U000F0F62" # moon-full
      - "\U000F0F63" # moon-last-quarter
      - "\U000F0F64" # moon-new
      - "\U000F0F65" # moon-waning-crescent
      - "\U000F0F66" # moon-waning-gibbous
      - "\U000F0F67" # moon-waxing-crescent
      - "\U000F0F68" # moon-waxing-gibbous

  - file: "materialdesignicons-webfont.ttf"
    id: font_icons_tiny
    size: 32
    glyphs:
      - "\U000F005E" # arrow-up-thick
      - "\U000F0046" # arrow-down-thick
      

display:
  - platform: ili9xxx
    model: S3BOX
    cs_pin: GPIO5
    dc_pin: GPIO4
    reset_pin: GPIO48
    id: lcd
    lambda: |-
      #define ICON_fan1     "\U000F1472"
      #define ICON_fan2     "\U000F1473"
      #define ICON_fan3     "\U000F1474"
      #define ICON_fanoff   "\U000F081D"
      #define ICON_lamp     "\U000F06B5"
      #define ICON_ceiling  "\U000F0769"   
      auto bg = Color(250, 250, 250);
      auto text = Color(250, 250, 250);
      auto border_color = Color(60, 60, 60);
      auto button_bg_on = Color(36, 180, 245);
      auto button_bg_off = Color(158, 158, 158);
      auto red = Color(255, 0, 0);
      auto green = Color(0, 255, 0);
      auto blue = Color(0, 0, 255);
      if (id(home_pressed).state == "off") {
      it.image(0,0, id(background_image), ImageAlign::TOP_LEFT);
      it.printf(160, 35, id(font_arial_big), TextAlign::CENTER, "%s", id(readable_ha_time).state.c_str());
      it.strftime(160, 90, id(font_medium), text, TextAlign::CENTER, "%A", id(time_ha).now());
      it.strftime(160, 125, id(font_medium), text, TextAlign::CENTER, "%m/%d/%Y", id(time_ha).now());
      if (id(backyard_temp).has_state()) {
        it.printf(0, 185, id(font_medium), text, TextAlign::LEFT, "Out: %.f°F", id(backyard_temp).state);
      }
      if (id(hitemp).has_state()) {
        it.printf(165, 185, id(font_small), text, TextAlign::LEFT, "Hi: %s°F", id(hitemp).state.c_str());
      }
      if (id(lowtemp).has_state()) {
        it.printf(165, 200, id(font_small), text, TextAlign::LEFT, "Lo: %s°F", id(lowtemp).state.c_str());
      }
      if (id(sunup).has_state()) {
        it.printf(235, 185, id(font_small), text, TextAlign::LEFT, "Rise: %s", id(sunup).state.c_str());
      }
      if (id(sundown).has_state()) {
        it.printf(235, 200, id(font_small), text, TextAlign::LEFT, "  Set: %s", id(sundown).state.c_str());
      }
      // moon phase
      auto moonIcon = moonToIcon(id(moon).state);      
      it.printf(20, 90, id(font_icons_med), TextAlign::LEFT, moonIcon.c_str());
      it.printf(40, 150, id(font_small), TextAlign::CENTER, id(moon_eng).state.c_str());
      // current condition icon
      auto dayTime = id(sun).state == "above_horizon";
      auto conditionIcon = conditionToIcon(id(w_condition).state, dayTime);
      it.printf(260, 90, id(font_icons_med), TextAlign::LEFT, conditionIcon.c_str());
      it.printf(280, 150, id(font_small), TextAlign::CENTER, id(condition_eng).state.c_str());
      it.printf(160, 230, id(font_small), TextAlign::CENTER, id(open_doors).state.c_str());
      } else {
      it.image(0,0, id(background_image), ImageAlign::TOP_LEFT);
      it.printf(40, 40, id(font_icons_med), TextAlign::CENTER, ICON_fan1);
      it.printf(40, 70, id(font_small), text, TextAlign::CENTER, "Low");
      it.printf(160, 40, id(font_icons_med), TextAlign::CENTER, ICON_fan2);
      it.printf(160, 70, id(font_small), text, TextAlign::CENTER, "Med");
      it.printf(280, 40, id(font_icons_med), TextAlign::CENTER, ICON_fan3);
      it.printf(280, 70, id(font_small), text, TextAlign::CENTER, "High");
      it.printf(160, 120, id(font_icons_med), TextAlign::CENTER, ICON_fanoff);
      it.printf(160, 150, id(font_small), text, TextAlign::CENTER, "Off"); 
      it.printf(90, 190, id(font_icons_med), TextAlign::CENTER, ICON_lamp);
      it.printf(90, 220, id(font_small), text, TextAlign::CENTER, "Lamp"); 
      it.printf(230, 190, id(font_icons_med), TextAlign::CENTER, ICON_ceiling); 
      it.printf(230, 220, id(font_small), text, TextAlign::CENTER, "Fan Light");

                  
      }

sensor:
  - platform: homeassistant
    id: backyard_temp
    entity_id: sensor.back_porch_fan_button_temperature
    internal: True

text_sensor:    
  - platform: homeassistant
    id: current_time
    entity_id: sensor.time
    internal: true
  - platform: homeassistant
    name: Readable_time
    entity_id: sensor.12h_time
    id: readable_ha_time   
  - platform: homeassistant
    id: lowtemp
    entity_id: sensor.temp_low
    internal: True
  - platform: homeassistant
    id: hitemp
    entity_id: sensor.temp_hi
    internal: True
  - platform: homeassistant
    id: sundown
    entity_id: sensor.sun_down
    internal: True
  - platform: homeassistant
    id: sunup
    entity_id: sensor.sun_up
    internal: True
  - platform: homeassistant
    id: moon
    entity_id: sensor.moon_phase
    internal: True
  - platform: homeassistant
    id: moon_eng
    entity_id: sensor.moon_phase_eng
    internal: True
  - platform: homeassistant
    id: condition_eng
    entity_id: sensor.condition_eng
    internal: True
  - platform: homeassistant
    entity_id: weather.blenheimdmz
    id: w_condition
    internal: true
  - platform: homeassistant
    entity_id: sun.sun
    id: sun
    internal: true
  - platform: homeassistant
    entity_id: sensor.open_doors
    id: open_doors
    internal: true
  - platform: homeassistant
    id: home_pressed
    entity_id: input_boolean.flexbox_home_pressed
    internal: True    

image:
  - file: "images/Graphic-Design-Background-320x240.jpg"
    id: background_image
    resize: 320x240
    type: rgba

time:
  - platform: homeassistant
    id: time_ha
    timezone: America/Chicago

psram:
  mode: octal
  speed: 80MHz
```


### Anything in the logs that might be useful for us?

```txt
Even with VERY_VERBOSE logging turned on (and removing the component: ERROR) I see nothing in the logs for touchscreen events.
```


### Additional information

_No response_

> could you try this as external_components: [esphome/esphome#5997](https://github.com/esphome/esphome/pull/5997)

I tried this:

```
external_components:
  - source: github://pr#5997
    components: [ tt21100 ]
    #components: [ touchscreen ]
```

And it fails to build:
```
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.4
Compiling .pioenvs/voice-assistant-1/src/esphome/components/tt21100/touchscreen/tt21100.o
Compiling .pioenvs/voice-assistant-1/src/main.o
src/esphome/components/tt21100/touchscreen/tt21100.cpp: In member function 'virtual void esphome::tt21100::TT21100Touchscreen::setup()':
src/esphome/components/tt21100/touchscreen/tt21100.cpp:67:42: error: 'class esphome::display::Display' has no member named 'get_native_width'; did you mean 'get_width'?
       this->x_raw_max_ = this->display_->get_native_width();
                                          ^~~~~~~~~~~~~~~~
                                          get_width
src/esphome/components/tt21100/touchscreen/tt21100.cpp:70:42: error: 'class esphome::display::Display' has no member named 'get_native_height'; did you mean 'get_height'?
       this->x_raw_max_ = this->display_->get_native_height();
                                          ^~~~~~~~~~~~~~~~~
                                          get_height
*** [.pioenvs/voice-assistant-1/src/esphome/components/tt21100/touchscreen/tt21100.o] Error 1
========================= [FAILED] Took 47.80 seconds =========================
```

--- a/esphome/components/touchscreen/touchscreen.cpp
+++ b/esphome/components/touchscreen/touchscreen.cpp
@@ -50,13 +50,15 @@ void Touchscreen::loop() {
       tp.second.x_prev = tp.second.x;
       tp.second.y_prev = tp.second.y;
     }
+    // The interrupt flag must be reset BEFORE calling update_touches, otherwise we might miss an interrupt that was
+    // triggered while we were reading touch data.
+    this->store_.touched = false;
     this->update_touches();
     if (this->skip_update_) {
       for (auto &tp : this->touches_) {
         tp.second.state &= ~STATE_RELEASING;
       }
     } else {
-      this->store_.touched = false;
       this->defer([this]() { this->send_touches_(); });
       if (this->touch_timeout_ > 0) {
         // Simulate a touch after <this->touch_timeout_> ms. This will reset any existing timeout operation.

